test/infrastructure/docker/api/v1beta2/devcluster_types.go

```diff
@@ -0,0 +1,194 @@
+	DevClusterDockerLoadBalancerAvailableV1Beta2Condition string = "LoadBalancerAvailable"
+
+	// DevClusterDockerLoadBalancerNotAvailableV1Beta2Reason surfaces when the container that implements
+	// the load balancer for a DevCluster's docker backend is not available.
+	DevClusterDockerLoadBalancerNotAvailableV1Beta2Reason = clusterv1.NotAvailableReason
+
+	// DevClusterDockerLoadBalancerAvailableV1Beta2Reason surfaces when the container that implements
+	// the load balancer for a DevCluster's docker backend is available.
+	DevClusterDockerLoadBalancerAvailableV1Beta2Reason = clusterv1.AvailableReason
+
+	// DevClusterDockerLoadBalancerDeletingV1Beta2Reason surfaces when the container that implements
+	// the load balancer for a DevCluster's docker backend is deleting.
+	DevClusterDockerLoadBalancerDeletingV1Beta2Reason = clusterv1.DeletingReason
+)
+
+// DevClusterSpec defines the desired state of the DevCluster infrastructure.
+type DevClusterSpec struct {
+	// controlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+	// +optional
+	ControlPlaneEndpoint APIEndpoint `json:"controlPlaneEndpoint"`
+
+	// backend defines backends for a DevCluster.
+	// +required
+	Backend DevClusterBackendSpec `json:"backend"`
+}
+
+// DevClusterBackendSpec defines backends for a DevCluster.
+type DevClusterBackendSpec struct {
+	// docker defines a backend for a DevCluster using docker containers.
+	// +optional
+	Docker *DockerClusterBackendSpec `json:"docker,omitempty"`
+
+	// inMemory defines a backend for a DevCluster that runs in memory.
+	// +optional
+	InMemory *InMemoryClusterBackendSpec `json:"inMemory,omitempty"`
+}
+
+// DockerClusterBackendSpec  defines a backend for a DevCluster using docker containers.
+type DockerClusterBackendSpec struct {
+	// failureDomains are usually not defined in the spec.
+	// The docker provider is special since failure domains don't mean anything in a local docker environment.
+	// Instead, the docker cluster controller will simply copy these into the Status and allow the Cluster API
+	// controllers to do what they will with the defined failure domains.
+	// +optional
+	FailureDomains clusterv1.FailureDomains `json:"failureDomains,omitempty"`
+
+	// loadBalancer allows defining configurations for the cluster load balancer.
+	// +optional
+	LoadBalancer DockerLoadBalancer `json:"loadBalancer,omitempty"`
+}
+
+// InMemoryClusterBackendSpec defines backend for a DevCluster that runs in memory.
+type InMemoryClusterBackendSpec struct{}
+
+// DevClusterStatus defines the observed state of the DevCluster.
+type DevClusterStatus struct {
+	// ready denotes that the dev cluster infrastructure is ready.
+	// +optional
+	Ready bool `json:"ready"`
+
+	// failureDomains don't mean much in CAPD since it's all local, but we can see how the rest of cluster API
+	// will use this if we populate it.
+	// +optional
+	FailureDomains clusterv1.FailureDomains `json:"failureDomains,omitempty"`
```

Comment:
So for now we are still implementing the v1beta1 contract with apiVersion v1beta2. That's okay. We'll change that in a follow-up. Please also declare the contract for v1beta2

test/infrastructure/docker/config/crd/kustomization.yaml:
```yaml
labels:
- pairs:
    cluster.x-k8s.io/v1beta1: v1beta1_v1beta2
```